FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -121,12 +127,148 @@ - (SEGPayload *) getNewPayloadWithFSURL:(SEGContext * _Nonnull)context {
             newPayload = nil;
             break;
     }
+
     return newPayload;
 }
 
+
+- (NSDictionary *) getSuffixedProps: (NSDictionary *)properties {
+    //TODO: Segment will crash and not allow props to have curcular dependency, but we should handle it here anyways
+    
+    NSMutableDictionary *props = [[NSMutableDictionary alloc] initWithCapacity:[properties count]];
+    for(NSString *key in properties){
+        NSString *suffix = @"";
+
+        // properties should always be a NSDictionary, but still check to make sure the value we get is an object
+        // Check Type Encoding: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html
+        const char* typeCode = @encode(typeof(props[key]));
+
+        if ([@"@" isEqualToString:@(typeCode)]) {
+            NSObject *obj = properties[key];
+            
+            // if it's a NSDictionary: recurrsively get props, and we can not take underscore in key for nested objects
+            // more info: https://help.fullstory.com/hc/en-us/articles/360020623234-FS-Recording-Client-API-Requirements
+            if ([obj isKindOfClass:[NSDictionary class]]) {
+                [props setValue:[self getSuffixedProps:properties[key]]
+                         forKey:[key stringByReplacingOccurrencesOfString:@"_" withString:@""]];
+            } else if ([obj isKindOfClass:[NSArray class]]) {
+                // array of dicts? array of strings/ints/numbers? yuck!
+                NSArray *arr = (NSArray *) obj;
+                NSDictionary *dict = [self getDictionaryFromArrayObject:arr withKey:key];
+                [props addEntriesFromDictionary:dict];
+            } else {
+                // if this is not a dictionary or array, then treat it like simple values, if data falls outside of these, we don't try to infer anything and just send as is to the server
+                suffix = [self getSuffixStringFromSimpleObject:obj];
+                [self appendToDictionary:props withKey:[key stringByAppendingString:suffix] andValue:properties[key]];
+            }
+        } else {
+            #ifdef DEBUG
+                NSAssert(FALSE, @"key `%@` is not an object can't be serialized for FS custom event.", key);
+            #else
+                NSLog(@"key `%@` is not an object can't be serialized for FS custom event.", key);
+                // if prod, then ignore this pops all together
+            #endif
+        }
+    }
+
+    return props;
+}
+
+- (NSString *) getSuffixStringFromSimpleObject: (NSObject *) obj {
```

Comment:
Super small i would tighten up the spaces on parameters, for example in the fullstory project this is how functions are written and most everywhere else:

```
+ (void)setDelegate:(id<FSDelegate>)delegate {
    fullstoryDelegate = delegate;
}
```